### ISSUE TITLE
createnotification runs only once for custom runner

### ISSUE THREAD

**Issue by sirpy** [Meteor-Reaction/Herald#48]
i've create a custom runner that only logs to console.
on the server when i issue Herald.createNotification either with an array of userIds or calling createNotification for each userId, the runner is ran only once. (it prints to the console only once - one user). subsequent calls to createnotification also doesnt do anything.

**Comment by Kestanous** [comment]
If i had to guess your console logs are running and being silenced by artwells:queue. Could you share your code or even share your repository? I am curious how it even logged once to be honest. Also are you looking to extend Herald or is this for private use?

**Comment by sirpy** [comment]
Well, i dont have a repository.
I'm trying to write a simple extension to incorporate raix:push (push notifications),
i'd be happy to share it once its working, so you can pack it as an extension.
here is the code so far:
**the extension:**

```javascript
var runner = {
  name: 'pushNotifications',
  where: ['server']
}
var pushConfiguration = null;
runner.run = function (notification, user) {
    if(!pushConfiguration)
    {
        Push.debug = this.debug || false;
        pushConfiguration = Push.Configure({
            gcm: {
                apiKey: this.gcm,
            },        
            production: true, // use production server or sandbox
        });      
    }
    console.log("Herald-PushNotifications: sending notification",notification,user);
    return true;
}

runner.check = function (notification, user) {
    console.log(notification);
  if (notification && (!notification.data || !notification.data.title))
    throw new Error('Herald-PushNotifications : notification "title" field must be defined');
  if(!(this.gcm || this.apn))  
      throw new Error('Herald-PushNotifications : atleast "apn" or "gcm" fields must be defined',this);
}

Herald.addRunner(runner);
```

**and the code for sending:
in startup:**
```javascript
Herald.settings.queueTimer = 1000;
    Herald.addCourier('cron', {
      media: {
        pushNotifications: {
          gcm: '***************',
          debug:true,
        }
      },
    });
```
**and the sending code:**
```javascript
_.each(workerIds,function(id){
                Herald.createNotification(id,{
                    courier:"cron",
                    data:{
                      from: 'xxxx,
                      title: 'xxxxxxxx',
                      text: '',
                      payload:{type:"approved",msgcnt:1},
                    },
                });
            });
```

**Comment by Kestanous** [comment]
Awesome, you are working on Push! That is on my todo list, raix:push has been looking good. I will review your code shortly but the I expect the queue your problem here. If your interested I can add you to the Herald Trello boards, not like there is anything their quite yet. We have been talking about reworking Herald for v2.0 soon. There are a few issue, especially server-side.

**Comment by sirpy** [comment]
anything i can do about the queue? if you have any updates, just add them here.
thanks

**Comment by Kestanous** [comment]
Ok I have a little time now. Are you just trying to confirm that it was sent or do you need the log for another reason? If you add artwells:queue to your app you can use Queue.log.find() to get the history.

**Comment by sirpy** [comment]
yes just want to confirm it reaches the run for each message. i think it installs artwelss:queue automatically.

**Comment by Kestanous** [comment]
It uses the queue internally but it does not imply it into your app. So if you want to get a the queue for review you need to add it to your app. Its just a namespace thing, I was not sure if users wanted the queue direct so a went the safer route.

**Comment by sirpy** [comment]
when i print to the console Queue.entries.find().fetch().
I get a really long list and they all show the same status. "command returned false" and "failed"
command: 'Meteor.call("heraldEscalate","ahvNpQF5CRB77BGyr", "pushNotifications")',
    created_at: Mon Jan 19 2015 08:04:00 GMT+0200 (IST),
    execute_after: Mon Jan 19 2015 08:04:00 GMT+0200 (IST),
    history: ' command returned false (Mon Jan 19 2015 08:04:27 GMT+0200 (IST));',
    priority: 5,
    status: 'failed',
    updated_at: Mon Jan 19 2015 08:04:27 GMT+0200 (IST) },

**Comment by Kestanous** [comment]
Tisk, debugging runners is a huge pain right now. Try the following code: https://gist.github.com/Kestanous/5359aa93bba5cb890a79
Put it in your app/package and run `runTestRunner()` on startup or when your ready.

Also be sure you have all the packages you need. If you are running this in a package remember that you must include them in your package.js.

**Comment by Kestanous** [comment]
Also does `Push.Configure` need to be run before every notification? Thats something I would force the users to configure.

**Comment by sirpy** [comment]
there's an IF that checks if pushconfiguration was already set. it doesnt run every time.

**Comment by Kestanous** [comment]
Ah sounds good. Also warning I did not run the code in that gist, I just copied what you gave me.

**Comment by Kestanous** [comment]
Oh now I see the if. Can you have multiple configurations though? What if a users has 5 push couriers, which config gets set?

**Comment by sirpy** [comment]
i guess it should be a different instance for each courier.
anyways i've decided to not use herald at the moment since i don't have the time to debug it.

**Comment by sirpy** [comment]
i found the issue to be in the configuration of Push.

**Comment by Kestanous** [comment]
Well I glad to here you found out the source but sorry to hear your not using herald. Let me know once you get push set up in your app and maybe we can build a simple runner around the Push.send function.

**Comment by sirpy** [comment]
once i found the issue, i kept using herald.
regarding the push notification this is a singleton, so i've removed the
configuration from the runner, this is something the user will have to
configure by his own, not related to the courier.

On Wed, Jan 21, 2015 at 3:36 PM, Sivli Kestanous <notifications@github.com>
wrote:

>  Well I glad to here you found out the source but sorry to hear your not
> using herald. Let me know once you get push set up in your app and maybe we
> can build a simple runner around the Push.send function.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Meteor-Reaction/Herald/issues/48#issuecomment-70837276>
> .
>



-- 
Hadar Rottenberg
050-7319093

**Comment by Daz2345** [comment]
Hi sirpy - looks like you may have achieved what I need, any chance you can share the final package that you came up with. Many Thanks

**Comment by sirpy** [comment]
I made no changes to the package.
I had an exception in the configuration stage that wasn't reported.
What are you trying to achieve?
